File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/diagnosiskeyssubmission/DiagnosisKeysSubmissionService.java
package uk.nhs.nhsx.diagnosiskeyssubmission;

import com.amazonaws.services.dynamodbv2.document.Item;
import org.apache.http.entity.ContentType;
import uk.nhs.nhsx.core.Jackson;
import uk.nhs.nhsx.core.aws.dynamodb.AwsDynamoClient;
import uk.nhs.nhsx.core.aws.s3.BucketName;
import uk.nhs.nhsx.core.aws.s3.ByteArraySource;
import uk.nhs.nhsx.core.aws.s3.Locator;
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider;
import uk.nhs.nhsx.core.aws.s3.S3Storage;
import uk.nhs.nhsx.core.events.Events;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.ClientTemporaryExposureKey;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.ClientTemporaryExposureKeysPayload;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKey;
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKeyPayload;
import uk.nhs.nhsx.keyfederation.InvalidRollingPeriod;
import uk.nhs.nhsx.keyfederation.InvalidRollingStartNumber;
import uk.nhs.nhsx.keyfederation.InvalidTemporaryExposureKey;
import uk.nhs.nhsx.keyfederation.InvalidTransmissionRiskLevel;
import uk.nhs.nhsx.virology.TestKit;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Base64;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Function;
import java.util.function.Supplier;

import static java.util.stream.Collectors.toList;

public class DiagnosisKeysSubmissionService {

    private static final String SUBMISSION_TOKENS_HASH_KEY = "diagnosisKeySubmissionToken";
    private static final int MAX_KEYS = 14;

    private final BucketName bucketName;
    private final S3Storage s3Storage;
    private final AwsDynamoClient awsDynamoClient;
    private final ObjectKeyNameProvider objectKeyNameProvider;
    private final String tableName;
    private final Supplier<Instant> clock;
    private final Events events;

    public DiagnosisKeysSubmissionService(S3Storage s3Storage,
                                          AwsDynamoClient awsDynamoClient,
                                          ObjectKeyNameProvider objectKeyNameProvider,
                                          String tableName,
                                          BucketName name,
                                          Supplier<Instant> clock,
                                          Events events) {
        this.bucketName = name;
        this.s3Storage = s3Storage;
        this.awsDynamoClient = awsDynamoClient;
        this.objectKeyNameProvider = objectKeyNameProvider;
        this.tableName = tableName;
        this.clock = clock;
        this.events = events;
    }

    public void acceptTemporaryExposureKeys(ClientTemporaryExposureKeysPayload payload) {
        allValidMaybe(payload)
            .ifPresent(this::acceptPayload);
    }

    private Optional<ClientTemporaryExposureKeysPayload> allValidMaybe(ClientTemporaryExposureKeysPayload payload) {
        if (payload.temporaryExposureKeys.size() > MAX_KEYS) {
            events.emit(getClass(), new TemporaryExposureKeysSubmissionOverflow(payload.temporaryExposureKeys.size(), MAX_KEYS));
            return Optional.empty();
        }

        var validKeys = payload.temporaryExposureKeys.stream().filter(this::isValidKey).collect(toList());
        var invalidKeysCount = payload.temporaryExposureKeys.size() - validKeys.size();

        events.emit(getClass(), new DownloadedTemporaryExposureKeys(validKeys.size(), invalidKeysCount));

        if (validKeys.size() > 0) {
            return Optional.of(new ClientTemporaryExposureKeysPayload(payload.diagnosisKeySubmissionToken, validKeys));
        }

        events.emit(getClass(), new EmptyTemporaryExposureKeys());
        return Optional.empty();
    }

    private Boolean isValidKey(ClientTemporaryExposureKey temporaryExposureKey) {
        return Optional.ofNullable(temporaryExposureKey)
            .filter(tek -> isKeyValid(tek.key))
            .filter(tek -> isRollingStartNumberValid(tek.rollingStartNumber, tek.rollingPeriod))
            .filter(tek -> isRollingPeriodValid(tek.rollingPeriod))
            .filter(tek -> isTransmissionRiskLevelValid(tek.transmissionRiskLevel))
            .isPresent();
    }

    private boolean isKeyValid(String key) {
        var isValid = key != null && isBase64EncodedAndLessThan32Bytes(key);
        if (!isValid) events.emit(getClass(), new InvalidTemporaryExposureKey(key));
        return isValid;
    }

    private boolean isRollingStartNumberValid(long rollingStartNumber, int rollingPeriod) {
        var now = clock.get();
        long TEN_MINUTES_INTERVAL_SECONDS = 600L;
        var currentInstant = now.getEpochSecond() / TEN_MINUTES_INTERVAL_SECONDS;
        var expiryPeriod = clock.get().minus(14, ChronoUnit.DAYS).getEpochSecond() / TEN_MINUTES_INTERVAL_SECONDS;
        var isValid = (rollingStartNumber + rollingPeriod) >= expiryPeriod && rollingStartNumber <= currentInstant;
        if (!isValid) {
            events.emit(getClass(), new InvalidRollingStartNumber(now, rollingStartNumber, rollingPeriod));
        }
        return isValid;
    }

    private boolean isRollingPeriodValid(int rollingPeriod) {
        boolean isValid = rollingPeriod > 0 && rollingPeriod <= 144;
        if(!isValid)  events.emit(getClass(), new InvalidRollingPeriod(rollingPeriod));
        return isValid;
    }

    private boolean isBase64EncodedAndLessThan32Bytes(String value) {
        try {
            byte[] key = Base64.getDecoder().decode(value);
            return key.length < 32;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    private boolean isTransmissionRiskLevelValid(int transmissionRiskLevel) {
        boolean isValid = transmissionRiskLevel >= 0 && transmissionRiskLevel <= 7;
        if(!isValid)  events.emit(getClass(), new InvalidTransmissionRiskLevel(transmissionRiskLevel));
        return isValid;
    }

    private void acceptPayload(ClientTemporaryExposureKeysPayload payload) {
        matchDiagnosisToken(payload.diagnosisKeySubmissionToken)
            .map(this::testkitFrom)
            .ifPresent(it -> storeKeysAndDeleteToken(it, payload));
    }

    private TestKit testkitFrom(Item item) {
        return Optional.ofNullable(item.get("testKit"))
            .map(Object::toString)
            .map(TestKit::valueOf)
            .orElse(TestKit.LAB_RESULT);
    }

    private Optional<Item> matchDiagnosisToken(UUID token) {
        Item item = awsDynamoClient.getItem(
            tableName,
            SUBMISSION_TOKENS_HASH_KEY,
            token.toString()
        );

        if (item == null) {
            events.emit(getClass(), new DiagnosisTokenNotFound(token));
        }

        return Optional.ofNullable(item);
    }

    private void storeKeysAndDeleteToken(TestKit testKit, ClientTemporaryExposureKeysPayload payload) {
        uploadToS3(testKit, payload);
        deleteToken(payload.diagnosisKeySubmissionToken);
    }

    private void uploadToS3(TestKit testKit, ClientTemporaryExposureKeysPayload payload) {
        var uploadPayload = convertToStoredModel(payload);
        var provider = new TestKitAwareObjectKeyNameProvider(objectKeyNameProvider, testKit);
        var objectKey = provider.generateObjectKeyName().append(".json");

        s3Storage.upload(
            Locator.of(bucketName, objectKey),
            ContentType.APPLICATION_JSON,
            ByteArraySource.fromUtf8String(Jackson.toJson(uploadPayload))
        );
    }

    private StoredTemporaryExposureKeyPayload convertToStoredModel(ClientTemporaryExposureKeysPayload payload) {
        return new StoredTemporaryExposureKeyPayload(
            payload.temporaryExposureKeys
                .stream()
                .map(buildStoredTemporaryExposureKey())
                .collect(toList())
        );
    }

    private Function<ClientTemporaryExposureKey, StoredTemporaryExposureKey> buildStoredTemporaryExposureKey() {
        return (ClientTemporaryExposureKey tek) -> new StoredTemporaryExposureKey
            (tek.key, tek.rollingStartNumber, tek.rollingPeriod, tek.transmissionRiskLevel, tek.daysSinceOnsetOfSymptoms);
    }

    private void deleteToken(UUID diagnosisKeySubmissionToken) {
        awsDynamoClient.deleteItem(
            tableName,
            SUBMISSION_TOKENS_HASH_KEY,
            diagnosisKeySubmissionToken.toString()
        );
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/diagnosiskeyssubmission/DiagnosisKeysSubmissionService.kt
package uk.nhs.nhsx.diagnosiskeyssubmission

import com.amazonaws.services.dynamodbv2.document.Item
import uk.nhs.nhsx.core.Clock
import uk.nhs.nhsx.core.ContentType.Companion.APPLICATION_JSON
import uk.nhs.nhsx.core.Json.toJson
import uk.nhs.nhsx.core.aws.dynamodb.AwsDynamoClient
import uk.nhs.nhsx.core.aws.s3.BucketName
import uk.nhs.nhsx.core.aws.s3.ByteArraySource.Companion.fromUtf8String
import uk.nhs.nhsx.core.aws.s3.Locator.Companion.of
import uk.nhs.nhsx.core.aws.s3.ObjectKeyNameProvider
import uk.nhs.nhsx.core.aws.s3.S3Storage
import uk.nhs.nhsx.core.events.Events
import uk.nhs.nhsx.diagnosiskeydist.agspec.RollingStartNumber.isRollingStartNumberValid
import uk.nhs.nhsx.diagnosiskeyssubmission.model.ClientTemporaryExposureKey
import uk.nhs.nhsx.diagnosiskeyssubmission.model.ClientTemporaryExposureKeysPayload
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKey
import uk.nhs.nhsx.diagnosiskeyssubmission.model.StoredTemporaryExposureKeyPayload
import uk.nhs.nhsx.diagnosiskeyssubmission.model.ValidatedTemporaryExposureKeysPayload
import uk.nhs.nhsx.keyfederation.InvalidRollingPeriod
import uk.nhs.nhsx.keyfederation.InvalidRollingStartNumber
import uk.nhs.nhsx.keyfederation.InvalidTemporaryExposureKey
import uk.nhs.nhsx.keyfederation.InvalidTransmissionRiskLevel
import uk.nhs.nhsx.domain.TestKit
import uk.nhs.nhsx.domain.TestKit.LAB_RESULT
import java.util.*

class DiagnosisKeysSubmissionService(
    private val s3Storage: S3Storage,
    private val awsDynamoClient: AwsDynamoClient,
    private val objectKeyNameProvider: ObjectKeyNameProvider,
    private val tableName: String,
    private val bucketName: BucketName,
    private val clock: Clock,
    private val events: Events
) {
    fun acceptTemporaryExposureKeys(payload: ClientTemporaryExposureKeysPayload) =
        allValidMaybe(payload).ifPresent(::acceptPayload)

    private fun allValidMaybe(payload: ClientTemporaryExposureKeysPayload) =
        when {
            payload.temporaryExposureKeys.size > MAX_KEYS -> {
                events(TemporaryExposureKeysSubmissionOverflow(payload.temporaryExposureKeys.size, MAX_KEYS))
                Optional.empty()
            }
            else -> {
                val validKeys = payload.temporaryExposureKeys.filterNotNull().filter(::isValidKey).map(::asStoredKey)

                val invalidKeysCount = payload.temporaryExposureKeys.size - validKeys.size
                events(DownloadedTemporaryExposureKeys(validKeys.size, invalidKeysCount))

                when {
                    validKeys.isNotEmpty() -> Optional.of(
                        ValidatedTemporaryExposureKeysPayload(
                            payload.diagnosisKeySubmissionToken,
                            validKeys
                        )
                    )
                    else -> {
                        events(EmptyTemporaryExposureKeys())
                        Optional.empty()
                    }
                }
            }
        }

    private fun isValidKey(temporaryExposureKey: ClientTemporaryExposureKey) =
        temporaryExposureKey
            .takeIf { isKeyValid(it.key) }
            ?.takeIf { isRollingStartNumberValid(it.rollingStartNumber.toLong(), it.rollingPeriod) }
            ?.takeIf { isRollingPeriodValid(it.rollingPeriod) }
            ?.takeIf { isTransmissionRiskLevelValid(it.transmissionRiskLevel) } != null

    private fun isKeyValid(key: String?): Boolean {
        val isValid = key != null && isBase64EncodedAndLessThan32Bytes(key)
        if (!isValid) events(InvalidTemporaryExposureKey(key))
        return isValid
    }

    private fun isRollingStartNumberValid(rollingStartNumber: Long, rollingPeriod: Int): Boolean {
        val now = clock()
        val isValid = isRollingStartNumberValid(
            { now },
            rollingStartNumber,
            rollingPeriod
        )
        if (!isValid) {
            events(InvalidRollingStartNumber(now, rollingStartNumber, rollingPeriod))
        }
        return isValid
    }

    private fun isRollingPeriodValid(rollingPeriod: Int): Boolean {
        val isValid = rollingPeriod in 1..144
        if (!isValid) events(InvalidRollingPeriod(rollingPeriod))
        return isValid
    }

    private fun isBase64EncodedAndLessThan32Bytes(value: String) = try {
        Base64.getDecoder().decode(value).size < 32
    } catch (e: IllegalArgumentException) {
        false
    }

    private fun isTransmissionRiskLevelValid(transmissionRiskLevel: Int): Boolean {
        val isValid = transmissionRiskLevel in 0..7
        if (!isValid) events(InvalidTransmissionRiskLevel(transmissionRiskLevel))
        return isValid
    }

    private fun acceptPayload(payload: ValidatedTemporaryExposureKeysPayload) {
        matchDiagnosisToken(payload.diagnosisKeySubmissionToken)
            .map(::testkitFrom)
            .ifPresent { storeKeysAndDeleteToken(it, payload) }
    }

    private fun testkitFrom(item: Item): TestKit = Optional.ofNullable(item["testKit"])
        .map(Any::toString)
        .map { TestKit.valueOf(it) }
        .orElse(LAB_RESULT)

    private fun matchDiagnosisToken(token: UUID): Optional<Item> {
        val item = awsDynamoClient.getItem(
            tableName,
            SUBMISSION_TOKENS_HASH_KEY,
            token.toString()
        )
        if (item == null) {
            events(DiagnosisTokenNotFound(token))
        }
        return Optional.ofNullable(item)
    }

    private fun storeKeysAndDeleteToken(testKit: TestKit, payload: ValidatedTemporaryExposureKeysPayload) {
        uploadToS3(testKit, payload)
        deleteToken(payload.diagnosisKeySubmissionToken)
    }

    private fun uploadToS3(testKit: TestKit, payload: ValidatedTemporaryExposureKeysPayload) {
        val uploadPayload = StoredTemporaryExposureKeyPayload(payload.temporaryExposureKeys)
        val provider = TestKitAwareObjectKeyNameProvider(objectKeyNameProvider, testKit)
        val objectKey = provider.generateObjectKeyName().append(".json")
        s3Storage.upload(
            of(bucketName, objectKey),
            APPLICATION_JSON,
            fromUtf8String(toJson(uploadPayload))
        )
    }

    private fun asStoredKey(it: ClientTemporaryExposureKey) =
        StoredTemporaryExposureKey(
            it.key!!,
            it.rollingStartNumber,
            it.rollingPeriod,
            it.transmissionRiskLevel,
            it.daysSinceOnsetOfSymptoms
        )

    private fun deleteToken(diagnosisKeySubmissionToken: UUID) = awsDynamoClient.deleteItem(
        tableName,
        SUBMISSION_TOKENS_HASH_KEY,
        diagnosisKeySubmissionToken.toString()
    )

    companion object {
        private const val SUBMISSION_TOKENS_HASH_KEY = "diagnosisKeySubmissionToken"
        private const val MAX_KEYS = 14
    }
}
